### PR TITLE
Avx512 support and spatial full distortion

### PR DESCRIPTION
**Pull request by lzhangnj**
Add AVX512 lib and spatial_full_distortion_kernel()

For spatial_full_distortion_kernel(), AVX512 is 0% to 40% faster than AVX2 version for width 64 and up.

Related to PR #414

### CODE REVIEW

**Comment by lzhangnj** [comment]
This PR blocks all other AVX512 works.

File: Source/Lib/Common/ASM_AVX512/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/C_DEFAULT/
+    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_SSE2/
+    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_SSSE3/
+    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_SSE4_1/
+    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_AVX2/
+    ${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_AVX512/)
+link_directories(${PROJECT_SOURCE_DIR}/Source/Lib/Common/ASM_SSSE3/)
+
+set(flags_to_test
+    -mavx2
```
Comment (1480c1) [inline Source/Lib/Common/ASM_AVX512/CMakeLists.txt]: Please check what cflags are needed for avx512 https://gcc.gnu.org/onlinedocs/gcc-9.1.0/gcc/x86-Options.html#x86-Options

File: Source/Lib/Common/ASM_AVX512/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+if(MSVC)
+    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} /arch:AVX2")
+endif()
+
+if(CMAKE_C_COMPILER_ID STREQUAL "Intel")
+    if(WIN32)
+        # Intel Windows (*Note - The Warning level /W0 should be made to /W4 at some point)
+        set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} /W0")
+    else()
+        set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -fPIC -static-intel -w")
```
Comment (1480c1) [inline Source/Lib/Common/ASM_AVX512/CMakeLists.txt]:
```suggestion
        set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -w")
```
static intel and fPIC is already set

**Comment by ttrigui** [comment]
@lzhangnj can you please look into the error

**Comment by lzhangnj** [comment]
LLVM 3.8 doesn't support AVX512. That's why some of these Travis CI builds failed when AVX512 code is activated by commit 747b3bf.

Any idea how to fix?

**Comment by 1480c1** [comment]
Either in cmake, check for clang, then version less than and exclude, or update the clang in Travis

**Comment by 1480c1** [comment]
Something like
``` cmake
if(CMAKE_C_COMPILER MATCHES "Clang" AND CMAKE_C_COMPILER_VERSION VERSION_LESS "Version")
...
```

**Comment by 1480c1** [comment]
I'm right now trying to see if we can use a fedora docker or the toolchain repo for Ubuntu in Travis to use a updated version of clang and gcc with ccache, @kylophone or @EwoutH any suggestions for possible travis changes?

**Comment by 1480c1** [comment]
According to https://docs.travis-ci.com/user/reference/xenial/, their xenial builds have llvm 7

**Comment by 1480c1** [comment]
Maybe `-mavx512f` https://clang.llvm.org/docs/ClangCommandLineReference.html#cmdoption-clang-mavx512f

**Comment by 1480c1** [comment]
Try removing clang from the apt in `.travis.yml`
``` diff
diff --git a/.travis.yml b/.travis.yml
index e259368..fe2e235 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -19,7 +19,6 @@ addons:
       - libgstreamer-plugins-base1.0-dev
       - libgstreamer1.0-dev
       - llvm
-      - clang
   homebrew:
     packages:
       - yasm
```

**Comment by 1480c1** [comment]
Remove llvm too and it will compile fine, for some reason, it was using the 3.8 clang header files even though clang 7 is the default clang installed on travis.

**Comment by lzhangnj** [comment]
Thank @1480c1 for fixing Travis CI builds!

I disabled AVX512 code to make this PR ready to merge to master. If you want to test AVX512 code accuracy and builds, please re-enable by commenting out "NON_AVX512_SUPPORT".

**Comment by hassount** [comment]
@mafonsoN 
any issues taking this in? it's a blocker for another PR

**Comment by mafonsoN** [comment]
> @mafonsoN
> any issues taking this in? it's a blocker for another PR

Looks fine to me. Go ahead.

**Pull request merged**